### PR TITLE
Empty IEntity

### PR DESCRIPTION
**Pull request by dberry-rcs**
!!!!!!!!!!!!!!!!!!!!!!    Tests are not passing, DO NOT MERGE to master     !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

I tried getting the library to support multiple ID types. The farthest I got was by deleteing ID from IEntity and changing GenerateNewID to return object. The code can now support _id, ID, or any [BsonId] property

Before I put more time in trying to fix the rest of the tests, I'd like to get your opinion on the direction. I tried several times to get IEntity working as IEntity<TId>, but at some point it hits a wall or you end up decorating everything with the type id. e.g. Cache<Book, string> DB.Collection<Book,string>, etc. I found it quite ugly.

The relationship code is not working well right now. ChildrenQueryable is not returning anything. add_child_to_many_relationship_with_ID is a good test to debug. If you could have a quick look I'd appreciate it.

### CODE REVIEW

**Comment by dj-nitehawk** [comment]
i like the direction you're going with the `object?` IDs.
also added some minor optimizations.
sooo... carry on i guess...
and let me know if you get stuck somewhere... i'll try my best to help ;-)

**Comment by dberry-rcs** [comment]
Understood about the breaking changes. As a user of the library, I am not looking for large refactorings on my projects.

**Comment by dberry-rcs** [comment]
@dj-nitehawk Is this a possibility for fixing the relations?

https://stackoverflow.com/questions/41093647/mongodb-join-on-id-field-from-string-to-objectid

if we do an .AppendStage, and do an $addfields to convert the id fields to strings and then always do a string comparison on the ids. This way, whether or not they are stored as ObjectIds, they would work.

It would require a minimum of mongodb 4.0

**Comment by dj-nitehawk** [comment]
might work... but i believe it would add a significant amount of overhead on the mongo server for something that is already a slowish process (lookups).

if we do a `$toString` on the local field, say in a db with millions of docs, the server now has to allocate millions of strings to compare against instead of what used to be an index scan. right?

**Comment by dberry-rcs** [comment]
@dj-nitehawk Looks like it is closer to passing. 1 error. JoinRecord needed a custom serializer that worked with object? I created a new attribute AsBsonId modeled after AsObjectId.  The relationships are now passing. Not sure what this latest error is.

File: MongoDB.Entities/Core/Cache.cs [hunk 1]
```diff
@@ -158,6 +158,7 @@ internal static IEnumerable<PropertyInfo> UpdatableProps(T entity)
     {
         var parameter = Expression.Parameter(typeof(T), "t");
         var property = Expression.Property(parameter, idProp);
-        return Expression.Lambda<Func<T, object?>>(property, parameter);
+        Expression conversion = Expression.Convert(property, typeof(object));
```
Comment (dj-nitehawk) [inline MongoDB.Entities/Core/Cache.cs:161]: looks like this conversion expression is not supported by mongo driver :-(

screen-record-1679888365

**Comment by dj-nitehawk** [comment]
can you try updating the project to the latest mongo driver?
[this issue](https://jira.mongodb.org/browse/CSHARP-4499?jql=project%20%3D%20CSHARP%20AND%20fixVersion%20%3D%202.19.1%20ORDER%20BY%20key%20ASC) might possibly be related.

**Comment by dberry-rcs** [comment]
No joy

**Comment by dberry-rcs** [comment]
@dj-nitehawk I saw that you pushed out a new version, so I updated the branch and pushed the PR. Still one error, but it is on Templates this time.

**Comment by dberry-rcs** [comment]
Looks like the ID field on the One class is not getting set correctly. I will keep working on it. So close.

**Comment by dberry-rcs** [comment]
@dj-nitehawk Finally it passes. Please take a look at One<T> to see the fix. I needed to store the T locally so that the ID field would have the correct type and serialization would work properly. If you are good with this I will start to look at the MultiDB support.

**Comment by dj-nitehawk** [comment]
great!

will review as soon as i can...

might be a few days though.

**Comment by dj-nitehawk** [comment]
so.... i think i've managed to solve it without having to store the `T` inside `One<>`. checkout the `dev` branch, my fix is in there.
i'll close this PR since it now conflicts with what i've done. if the `dev` branch looks ok, would it be possible for you to add a few tests that cover usage with a couple of different `ID` property types which people might wanna be using it with? you know guid/int/etc...

**Comment by dberry-rcs** [comment]
Will do.

**Comment by dj-nitehawk** [comment]
> Will do.

hey mate!
have you had a chance to look at the `dev` branch yet?
no rush ;-)

**Comment by dberry-rcs** [comment]
> 

I have. I think your changes look good, but wanted to make sure they are well tested. I tried refactoring the tests so that we can rerun all the tests for String, Long, ObjectID, and GUID ids. Unfortunately, you cannot have tests that take a generic, so I am trying to figure out a clean way to do this without have four sets of almost identical tests.

**Pull request closed without merging**